### PR TITLE
Java: Add flow sources and steps for RabbitMQ and JMS

### PR DESCRIPTION
**Pull request by artem-smotrakov**
This pull request was slpit from https://github.com/github/codeql/pull/8501 as [suggested](https://github.com/github/codeql/pull/8501#issuecomment-1073817544) by @smowton.

In addition, I've added flow sources and steps for JMS API versions 1 and 2 as [suggested](https://github.com/github/securitylab/issues/556#issuecomment-1073775894) by @pwntester. The model for the JMS API applis to all JMS implementations such as ActiveMQ and others. However, there may be implementation-specific flow sources and steps that are not covered here. I'd suggest to start with the API and then cover specific implementations if necessary.

Also, added a few flow steps for methods in `DataInput` and `ObjectInput` that read byte arrays.

This PR is related to https://github.com/github/securitylab/issues/666

### CODE REVIEW

**Comment by pwntester** [comment]
thanks @artem-smotrakov would it make sense to also add deserialization sinks to `ObjectMessage.getObject()` API as per https://www.blackhat.com/docs/us-16/materials/us-16-Kaiser-Pwning-Your-Java-Messaging-With-Deserialization-Vulnerabilities.pdf

File: java/ql/lib/semmle/code/java/frameworks/JMS.qll [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        // if a destination is tainted, then it returns tainted data
+        "javax.jms;Queue;true;getQueueName;();;Argument[-1];ReturnValue;taint",
+        "javax.jms;Queue;true;toString;();;Argument[-1];ReturnValue;taint",
+        "javax.jms;Topic;true;getTopicName;();;Argument[-1];ReturnValue;taint",
+        "javax.jms;Topic;true;toString;();;Argument[-1];ReturnValue;taint",
+      ]
+  }
+}
+
+/** Defines additional sources of taited data in JMS 2. */
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/frameworks/JMS.qll:94]:
```suggestion
/** Defines additional sources of tainted data in JMS 2. */
```

File: java/ql/src/change-notes/2022-04-17-jms.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+category: majorAnalysis
```
Comment (smowton) [inline java/ql/src/change-notes/2022-04-17-jms.md:2]:
```suggestion
category: minorAnalysis
```

File: java/ql/test/library-tests/frameworks/jms/FlowTest.ql [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  }
+}
+
+class JmsFlowTest extends InlineExpectationsTest {
+  JmsFlowTest() { this = "JmsFlowTest" }
+
+  override string getARelevantTag() { result = "detected" }
+
+  override predicate hasActualResult(Location location, string element, string tag, string value) {
+    tag = "detected" and
```
Comment (smowton) [inline java/ql/test/library-tests/frameworks/jms/FlowTest.ql:23]:
```suggestion
    tag = "tainted" and
```
(change correspondingly in the tests)

File: java/ql/lib/semmle/code/java/dataflow/FlowSteps.qll [hunk 4]
```diff
@@ -21,6 +21,7 @@ private module Frameworks {
   private import semmle.code.java.frameworks.guava.Guava
   private import semmle.code.java.frameworks.apache.Lang
   private import semmle.code.java.frameworks.ApacheHttp
+  private import semmle.code.java.frameworks.RabbitMQ
```
Comment (smowton) [inline java/ql/lib/semmle/code/java/dataflow/FlowSteps.qll:24]:
```suggestion
```

**Review by smowton — COMMENTED** [review]
Haven't reviewed the APIs in general to find additional methods that maybe ought to be modelled, but the changes here seem uncontroverisal

**Comment by artem-smotrakov** [comment]
Hi @smowton Thanks for the comments! I've addressed them.

**Comment by artem-smotrakov** [comment]
> thanks @artem-smotrakov would it make sense to also add deserialization sinks to `ObjectMessage.getObject()` API as per https://www.blackhat.com/docs/us-16/materials/us-16-Kaiser-Pwning-Your-Java-Messaging-With-Deserialization-Vulnerabilities.pdf

@pwntester `ObjectMessage.getObject()` is an interface method. It depends on implementation if and how it does deserialization. By adding this sink, we'd assume that all implementations do some kind of unsafe deserialization here.

The author of the paper you mentioned says

> All ObjectMessage implementations I looked at were deserializing from untrusted input without any validation 

Maybe it is a safe assumption. Let me know if this assumption is okay. I don't mind adding this sink though.

**Comment by pwntester** [comment]
> Maybe it is a safe assumption. Let me know if this assumption is okay. I don't mind adding this sink though.

I think that after the talk some mitgations were put in place such as the [ActiveMQ one](https://activemq.apache.org/security-advisories.data/CVE-2015-5254-announcement.txthttps://activemq.apache.org/security-advisories.data/CVE-2015-5254-announcement.txt) so it may require a little testing

**Comment by artem-smotrakov** [comment]
@smowton Unfortunately, I can't see why the checks failed. Could you please check what happened?

**Comment by atorralba** [comment]
### First error

File "ql/java/ql/test/stubs/rabbitmq-4.12.0/com/rabbitmq/client/DefaultConsumer.java" contains a non-ASCII character at the location marked with `|` in:
```
        
    }

    @Override
    public void handleDelivery|
```
  
File "ql/java/ql/test/stubs/rabbitmq-4.12.0/com/rabbitmq/client/Consumer.java" contains a non-ASCII character at the location marked with `|` in:
```
ption;;
  
public interface Consumer {
    void handleDelivery|
```

### Second error

```
--- expected
+++ actual
@@ -1,1 +1,2 @@
-
+| TestIO.java:97:14:97:58 | readLine(...) | Fixed missing result:numTaintFlow=1 |
+| TestIO.java:139:14:139:64 | readUTF(...) | Fixed missing result:numTaintFlow=1 |
Error: [7/8] [114/134 comp 13s eval 6.2s] FAILED(RESULT) ql/java/ql/test/library-tests/frameworks/guava/handwritten/flow.ql
```

### Third error

```
Error: The following CodeQL elements are lacking documentation:
semmle/code/java/frameworks/JMS.qll  JMS  file
```

**Comment by artem-smotrakov** [comment]
Thanks for the info @atorralba ! I hope the errors are now fixed.

**Review by smowton — DISMISSED** [review]

File: java/ql/lib/semmle/code/java/frameworks/JMS.qll [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        "javax.jms;JMSConsumer;true;receiveNoWait;();;ReturnValue;remote",
+        "javax.jms;JMSConsumer;true;receiveBodyNoWait;();;ReturnValue;remote",
+      ]
+  }
+}
+
+/** Defines additional taint propagation steps in JMS 2. */
+private class Jms2FlowStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row = ["javax.jms;Message;true;getBody;();;Argument[-1];ReturnValue;taint",]
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/frameworks/JMS.qll]: Calling this a "vulnerability" is an interesting wording 😄

**Review by smowton — APPROVED** [review]
@aschackmull is this ok with you with the import fixed as asked?

**Comment by aschackmull** [comment]
> @aschackmull is this ok with you with the import fixed as asked?

Change note possibly needs to be in `lib` rather than `src`. Otherwise yes, no further comments from me.

**Review by smowton — APPROVED** [review]

**Comment by artem-smotrakov** [comment]
Thanks for the review and addressing the comments @smowton !

**Pull request merged**